### PR TITLE
pallet-session: Migrate the historical part to the new pallet macro

### PR DESCRIPTION
**Pull request by koushiro**
Part of #7882

Migrate historical part of `pallet-session` to the new pallet attribute macro.

## :warning:  Breaking Change :warning:

From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines

> storages now use PalletInfo for module_prefix instead of the one given to decl_storage: Thus any use of this pallet in construct_runtime! should be careful to update name in order not to break storage or to upgrade storage (moreover for instantiable pallet). If pallet is published, make sure to warn about this breaking change.

So users of the pallet must be careful about the name they used in `construct_runtime!`. Hence the `runtime-migration` label, which might not be needed depending on the configuration of the pallet.

polkadot, kusama, westend and rococo use `Historical` as pallet name (but the current module_prefix is `Session`), thus need to migrate the storages.

---

polkadot companion: paritytech/polkadot#3949

Polkadot address: 15XuanNimo5951s3RjFTPX1AvYVnCwfr3SDhb3AV4fQF3LpK

### CODE REVIEW

File: frame/session/src/historical/mod.rs [hunk 1]
```diff
@@ -26,62 +26,84 @@
...
+
+	/// Deprecated.
+	#[pallet::storage]
+	pub type CachedObsolete<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		SessionIndex,
+		Vec<(T::ValidatorId, T::FullIdentification)>,
+		OptionQuery,
+	>;
```
Comment (koushiro) [inline frame/session/src/historical/mod.rs:103]: I'm not sure if I need to delete this obsolete storage

File: frame/staking/src/mock.rs [hunk 2]
```diff
@@ -96,6 +96,7 @@ frame_support::construct_runtime!(
 		Balances: pallet_balances::{Pallet, Call, Storage, Config<T>, Event<T>},
 		Staking: pallet_staking::{Pallet, Call, Config<T>, Storage, Event<T>},
 		Session: pallet_session::{Pallet, Call, Storage, Event, Config<T>},
+		Historical: pallet_session::historical::{Pallet, Storage},
```
Comment (kianenigma) [inline frame/staking/src/mock.rs:99]: Why?

Comment (thiolliere) [reply frame/staking/src/mock.rs:99]: I think it now needs to have its name in PalletInfo, for the storages, so it needs to be in construct_runtime

**Review by kianenigma — APPROVED** [review]
LGTM!

**Comment by emostov** [comment]
bot merge

**Comment by thiolliere** [comment]
bot merge

**Comment by thiolliere** [comment]
bot merge

**Comment by thiolliere** [comment]
if the bot give up on us, we will need another approval on the companion in order to merge manually the companion on polkadot

**Review by emostov — APPROVED** [review]

**Comment by emostov** [comment]
bot merge

**Comment by thiolliere** [comment]
bot merge force

**Pull request merged**